File: src/struct/MinHeap.java
package struct;

import annotation.HotCode;
import primitive.IntFloat;

import nio.StringLookup;
/**
 * Zero-GC Off-Heap Min-Heap implementation.
 * Wraps an IntFloat array pointer.
 * Int = item (e.g. entity ID), Float = priority/cost.
 * Uses 1-based indexing for optimized heap math.
 * Index 0 stores the current size of the heap in the Int part.
 */
@HotCode
public final class MinHeap {

    private MinHeap() {}

    @HotCode
    public static long allocate(int capacity) {
        // allocate capacity + 1 (for 1-based indexing and storing size at index 0)
        long ptr = IntFloat.allocateArray(capacity + 1);
        IntFloat.setUnsafe(ptr, 0, 0, 0f); // size = 0
        return ptr;
    }

    @HotCode
    public static void free(long ptr) {
        IntFloat.free(ptr);
    }

    @HotCode
    public static int size(long ptr) {
        return IntFloat.unsafeGetIntPart(ptr, 0);
    }

    @HotCode
    public static int capacity(long ptr) {
        return IntFloat.length(ptr) - 1;
    }

    @HotCode
    public static boolean isEmpty(long ptr) {
        return size(ptr) == 0;
    }

    @HotCode
    public static void push(long ptr, int item, float priority) {
        int sz = size(ptr);
        sz++;
        
        int cap = capacity(ptr);
        if (sz > cap) {
            throw new IllegalStateException(StringLookup.getJavaString(389) + cap);
        }
        
        IntFloat.setUnsafe(ptr, 0, sz, 0f); // update size
        IntFloat.setUnsafe(ptr, sz, item, priority); // place at end
        siftUp(ptr, sz); // sift up to correct position
    }

    @HotCode
    public static int popItem(long ptr) {
        int sz = size(ptr);
        if (sz == 0) throw new IllegalStateException(StringLookup.getJavaString(390));
        
        // Root is always at index 1
        int result = IntFloat.unsafeGetIntPart(ptr, 1);
        
        // Move last element to root
        int lastItem = IntFloat.unsafeGetIntPart(ptr, sz);
        float lastPrio = IntFloat.unsafeGetFloatPart(ptr, sz);
        IntFloat.setUnsafe(ptr, 1, lastItem, lastPrio);
        
        sz--;
        IntFloat.setUnsafe(ptr, 0, sz, 0f);
        
        if (sz > 0) {
            siftDown(ptr, 1);
        }
        
        return result;
    }

    @HotCode
    private static void siftUp(long ptr, int index) {
        int item = IntFloat.unsafeGetIntPart(ptr, index);
        float prio = IntFloat.unsafeGetFloatPart(ptr, index);
        
        while (index > 1) {
            int parent = index / 2;
            float parentPrio = IntFloat.unsafeGetFloatPart(ptr, parent);
            
            if (prio >= parentPrio) {
                break;
            }
            
            // move parent down
            int parentItem = IntFloat.unsafeGetIntPart(ptr, parent);
            IntFloat.setUnsafe(ptr, index, parentItem, parentPrio);
            index = parent;
        }
        IntFloat.setUnsafe(ptr, index, item, prio);
    }

    @HotCode
    private static void siftDown(long ptr, int index) {
        int sz = size(ptr);
        int item = IntFloat.unsafeGetIntPart(ptr, index);
        float prio = IntFloat.unsafeGetFloatPart(ptr, index);
        
        int half = sz / 2;
        while (index <= half) {
            int left = index * 2;
            int right = left + 1;
            
            int bestChild = left;
            float bestPrio = IntFloat.unsafeGetFloatPart(ptr, left);
            
            if (right <= sz) {
                float rightPrio = IntFloat.unsafeGetFloatPart(ptr, right);
                if (rightPrio < bestPrio) {
                    bestChild = right;
                    bestPrio = rightPrio;
                }
            }
            
            if (prio <= bestPrio) {
                break;
            }
            
            // move child up
            int childItem = IntFloat.unsafeGetIntPart(ptr, bestChild);
            IntFloat.setUnsafe(ptr, index, childItem, bestPrio);
            index = bestChild;
        }
        IntFloat.setUnsafe(ptr, index, item, prio);
    }
}

File: src/struct/minheap.c
#include "struct/minheap.h"

#include <string.h>

#include "nio/mem.h"
#include "oop/type.h"

// minheap.c — MinHeap port (Legacy: struct/MinHeap.java). 1-based array, sift
// up on push, sift down on pop, strictly lower priority rises to the root.

static void sift_up(MinHeap *heap, size_t index) {
    int32_t item = (*heap).items[index];
    float prio = (*heap).priorities[index];
    while (index > 1) {
        size_t parent = index / 2;
        float parent_prio = (*heap).priorities[parent];
        if (prio >= parent_prio)
            break;
        (*heap).items[index] = (*heap).items[parent];
        (*heap).priorities[index] = parent_prio;
        index = parent;
    }
    (*heap).items[index] = item;
    (*heap).priorities[index] = prio;
}

static void sift_down(MinHeap *heap, size_t index) {
    size_t size = (size_t)(*heap).size;
    int32_t item = (*heap).items[index];
    float prio = (*heap).priorities[index];
    size_t half = size / 2;
    while (index <= half) {
        size_t left = index * 2;
        size_t right = left + 1;

        size_t best = left;
        float best_prio = (*heap).priorities[left];
        if (right <= size) {
            float right_prio = (*heap).priorities[right];
            if (right_prio < best_prio) {
                best = right;
                best_prio = right_prio;
            }
        }
        if (prio <= best_prio)
            break;
        (*heap).items[index] = (*heap).items[best];
        (*heap).priorities[index] = best_prio;
        index = best;
    }
    (*heap).items[index] = item;
    (*heap).priorities[index] = prio;
}

MinHeap *MinHeap_allocate(size_t capacity) {
    size_t slots = capacity + 1;
    MinHeap *heap = (MinHeap *)Memory_alloc(TYPE_MIN_HEAP, sizeof(MinHeap));
    if (!heap) return nullptr;

    (*heap).size = 0;
    (*heap).capacity = (int32_t)capacity;
    (*heap).items = (int32_t *)Memory_alloc(TYPE_INT_ARRAY, slots * sizeof(int32_t));
    (*heap).priorities = (float *)Memory_alloc(TYPE_FLOAT_ARRAY, slots * sizeof(float));
    if (!(*heap).items || !(*heap).priorities) {
        Memory_free((*heap).items);
        Memory_free((*heap).priorities);
        Memory_free(heap);
        return nullptr;
    }
    return heap;
}

void MinHeap_free(MinHeap *heap) {
    if (!heap) return;
    Memory_free((*heap).items);
    Memory_free((*heap).priorities);
    Memory_free(heap);
}

size_t MinHeap_size(MinHeap *heap) {
    if (!heap) return 0;
    return (size_t)(*heap).size;
}

size_t MinHeap_capacity(MinHeap *heap) {
    if (!heap) return 0;
    return (size_t)(*heap).capacity;
}

bool MinHeap_isEmpty(MinHeap *heap) {
    return MinHeap_size(heap) == 0;
}

int MinHeap_push(MinHeap *heap, int32_t item, float priority) {
    if (!heap) return 0;
    if ((*heap).size >= (*heap).capacity)
        return 0;
    (*heap).size++;
    size_t index = (size_t)(*heap).size;
    (*heap).items[index] = item;
    (*heap).priorities[index] = priority;
    sift_up(heap, index);
    return 1;
}

int32_t MinHeap_popItem(MinHeap *heap) {
    if (!heap) return 0;
    if ((*heap).size == 0)
        return 0;
    int32_t result = (*heap).items[1];
    size_t size = (size_t)(*heap).size;
    (*heap).items[1] = (*heap).items[size];
    (*heap).priorities[1] = (*heap).priorities[size];
    (*heap).size--;
    if ((*heap).size > 0)
        sift_down(heap, 1);
    return result;
}

int32_t MinHeap_peekItem(MinHeap *heap) {
    if (!heap) return 0;
    if ((*heap).size == 0)
        return 0;
    return (*heap).items[1];
}